jax/experimental/pjit.py

```diff
@@ -457,11 +457,14 @@ def _pjit_translation_rule(c, axis_env, in_nodes, name_stack, backend, name,
 
 
 def _pjit_batcher(insert_axis,
+                  frame,
                   vals_in, dims_in,
                   axis_name, main_type,
                   jaxpr, in_axis_resources, out_axis_resources,
                   resource_env, donated_invars, name):
-  axis_size, = {x.shape[d] for x, d in zip(vals_in, dims_in) if d is not batching.not_mapped}
+  axis_size = frame.axis_size
+  axis_name = frame.name
+  main_type = frame.main_trace.trace_type
```

Comment:
Honestly I think it would be better to make `collective_rules` have the same signature as `initial_style_batchers` used to have.  Taking in `frame` causes a lot of code duplication, as can be seen here, and exposes lots of implementation details from `core`. It would be better to do all of those unpacks in `batching.py`.

Reply:
Looks good to me 🤷‍♂️